## Import Libraries {-}

In [14]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy import stats

## Read the Dataset

In [19]:
Data=pd.read_csv('/Users/harsha/Desktop/myProjects/Fall2020-Local/ML/CS536FinalProject/ML-FinalProject/Datasets/Test_data.csv', encoding = 'ISO-8859-1')
selFeat=['x1', 'y1', 'z1','x2', 'y2', 'z2', 'x3', 'y3', 'z3', 'x4', 'y4', 'z4']
Data=Data[selFeat]
Data_Array=Data.to_numpy()
BinomArray = np.random.binomial(1, 0.5, Xarr_val.shape)
indices=np.where(BinomArray==0)
TrueValues=Data_Array[indices]
Data_transformed=Xarr_val.astype('float')
Data_transformed[indices]=np.nan
Xtest=Data_transformed.copy()

## Nan added Dataset

In [20]:
Data_transformed

array([[  nan,   nan,  -82., ..., -179.,   nan, -178.],
       [ -14.,  107., -129., ..., -187.,   nan,   nan],
       [  nan,   69.,   nan, ...,   nan,   nan,   nan],
       ...,
       [  nan,   nan,  -55., ...,   nan, -109.,   nan],
       [   0.,   94., -149., ..., -216.,   nan, -156.],
       [  nan,   nan,  -96., ..., -181.,  -89.,   nan]])

## Mode as Basic Completion Agent

In [24]:
Data_Mode_Imputation=Data_transformed.copy()

In [26]:
col_mode = stats.mode(Data_Array)
#Find indices that you need to replace
inds = np.where(np.isnan(Data_Mode_Imputation))
#Place column modes in the indices. Align the arrays using take
Data_Mode_Imputation[inds] = np.take(col_mode[0], inds[1])
# print(X_train)
Data_Mode_Imputation

array([[  -1.,   95.,  -82., ..., -179.,  -94., -178.],
       [ -14.,  107., -129., ..., -187.,  -94., -163.],
       [  -1.,   69.,  -98., ..., -164.,  -94., -163.],
       ...,
       [  -1.,   95.,  -55., ..., -164., -109., -163.],
       [   0.,   94., -149., ..., -216.,  -94., -156.],
       [  -1.,   95.,  -96., ..., -181.,  -89., -163.]])

## Mean as Basic Completion Agent

In [28]:
Data_Mean_Imputation=Data_transformed.copy()

In [29]:
col_mean = np.nanmean(Xtest, axis=0)
# print(col_mean)
#Find indices that you need to replace
inds = np.where(np.isnan(Xtest))
#Place column means in the indices. Align the arrays using take
Data_Mean_Imputation[inds] = np.take(col_mean, inds[1])
# print(X_train)
Data_Mean_Imputation

array([[  -6.74788443,   88.26345372,  -82.        , ..., -179.        ,
         -92.52789256, -178.        ],
       [ -14.        ,  107.        , -129.        , ..., -187.        ,
         -92.52789256, -159.56642212],
       [  -6.74788443,   69.        ,  -93.00722848, ..., -167.19905814,
         -92.52789256, -159.56642212],
       ...,
       [  -6.74788443,   88.26345372,  -55.        , ..., -167.19905814,
        -109.        , -159.56642212],
       [   0.        ,   94.        , -149.        , ..., -216.        ,
         -92.52789256, -156.        ],
       [  -6.74788443,   88.26345372,  -96.        , ..., -181.        ,
         -89.        , -159.56642212]])

## Score between True Values & Feature Means

In [30]:
Means=Data_Mean_Imputation[indices]

In [31]:
score = (((TrueValues - Means) ** 2).sum() / ((TrueValues - TrueValues.mean()) ** 2).sum())
score

0.5542645175737081

## Score between True Values & Feature Modes

In [32]:
Modes=Data_Mode_Imputation[indices]

In [33]:
score = (((TrueValues - Modes) ** 2).sum() / ((TrueValues - TrueValues.mean()) ** 2).sum())
score

3.199443618413495

## Observations {-}
- Mean value imputation has a score of 0.55 and Mode value imputation scores 3.1 as we see both are not good imputation models compared to all other models we built for this problem statement